### PR TITLE
 Transfer web proxy steps to extension-selenium

### PR DESCRIPTION
**Pull request by VitaliaPiliuhina**

### CODE REVIEW

File: vividus-extension-selenium/build.gradle [hunk 1]
```diff
@@ -32,6 +33,11 @@ project.description = 'Vividus extension for selenium'
      implementation(group: 'org.apache.commons', name: 'commons-lang3', version: versions.commonsLang3)
      implementation(group: 'org.springframework', name: 'spring-core', version: versions.spring)
 
+     // browserup-proxy depends on both netty-all and netty-core, so we need to align versions of all netty dependencies
+     // https://github.com/browserup/browserup-proxy/blob/26714e26fae728733a13a0f332bd0c6dcaa78de5/browserup-proxy-core/build.gradle#L60-L76
+     implementation(group: 'io.netty', name: 'netty-all', version: "${nettyVersion}")
```
Comment (abudevich) [inline vividus-extension-selenium/build.gradle:38]: Why don't we write version as for `seleniumVersion` or `browserupProxyVersion` ?

File: docs/modules/plugins/partials/proxy-steps.adoc [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+----
+Given I am on a page with the URL 'https://www.google.com/search?q=vividus'
+When I capture HTTP GET or POST request with URL pattern `.*/search.*=vividus` and save URL to scenario variable `URL`
+Then `${URL}` is equal to `https://www.google.com/search?q=vividus`
+----
+
+.Validate the URL query of the matching HTTP request
+[source,gherkin]
+----
+Given I am on a page with the URL 'https://www.google.com/search?q=vividus'
```
Comment (valfirst) [inline docs/modules/plugins/partials/proxy-steps.adoc:78]: these examples are not applicable for mobile apps

File: docs/modules/plugins/pages/plugin-mobile-app.adoc [hunk 3]
```diff
@@ -744,3 +746,5 @@ When I select next value with `0.1` offset in picker wheel located `xpath(//XCUI
 
 
 include::partial$generic-ui-steps.adoc[]
+
+include::partial$proxy-steps.adoc[]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-mobile-app.adoc:750]: we should put some extra info, that proxy is not supported by Appium and custom solutions are needed

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — COMMENTED** [review]
Also please make sure the corresponding unit tests are moved as well

File: docs/modules/plugins/partials/proxy-steps.adoc (hunk 2 above)
Comment (ikalinin1) [reply docs/modules/plugins/partials/proxy-steps.adoc:78]: {givent-step}

File: docs/modules/plugins/pages/plugin-mobile-app.adoc [hunk 4]
```diff
@@ -161,9 +163,13 @@ a|Defines a destination folder on a device for a file being uploaded
 !`/sdcard/Pictures`
 
 !===
-
+include::partial$proxy-properties.adoc[]
 |===
 
+include::partial$proxy-meta-tags.adoc[]
+
+IMPORTANT: For device configuration, we need to perform the following steps: install CA certificate(download CA certificate from a browserup repository), configure device proxy, and configure the test app to allow proxy in cases of necessity.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-mobile-app.adoc:171]: please add a link to CA certificate to download

Comment (valfirst) [inline docs/modules/plugins/pages/plugin-mobile-app.adoc:171]: could you please elaborate on "configure device proxy"?

File: docs/modules/plugins/pages/plugin-mobile-app.adoc [hunk 5]
```diff
@@ -1,5 +1,7 @@
 :grid-default-hostname: empty
 :keys: https://github.com/appium/appium-xcuitest-driver#mobile-pressbutton[iOS keys] or https://appium.github.io/java-client/io/appium/java_client/android/nativekey/AndroidKey.html[Android keys]
+:proxy: This step requires proxy to be turned on. It can be done in properties or by switching on @proxy meta tag at the story level.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-mobile-app.adoc:3]: can it be a part of `proxy-steps.adoc`?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-mobile-app.adoc (hunk 4 above)
Comment (VitaliaPiliuhina) [reply docs/modules/plugins/pages/plugin-mobile-app.adoc:171]: I mean the following steps: 
- go to the device’s Settings and Wifi configuration 
- press on the network you’re going to use
- tap a Proxy field and select Manual  
- add a proxy hostname and a proxy port

File: docs/modules/plugins/partials/proxy-properties.adoc [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+|false
+|Enabled proxy for the run
+
+|proxy.host
+|any host name resolving on the machine
+|empty
+|Overrides the host which will be used by proxy
+
+|proxy.ports
+|10000-10005
```
Comment (uarlouski) [inline docs/modules/plugins/partials/proxy-properties.adoc:16]: can't it be any port from not reserved/registered range i.e. 49152 to 65535?

Comment (uarlouski) [reply docs/modules/plugins/partials/proxy-properties.adoc:16]: just wondering why 10000-10005

**Review by valfirst — CHANGES_REQUESTED** [review]
Please resolve conflicts at first

File: .github/workflows/gradle.yml [hunk 7]
```diff
@@ -282,6 +282,26 @@ jobs:
...
+                                                  -Pvividus.allure.executor.build-order=${GITHUB_RUN_ID} \
+                                                  -Pvividus.allure.executor.build-name="SauceLabs iOS Proxy Tests ${GITHUB_RUN_ID}" \
+                                                  -Pvividus.allure.executor.build-url=https://github.com/vividus-framework/vividus/actions/runs/${GITHUB_RUN_ID} \
+                                                  -Pvividus.allure.executor.report-url=https://github.com/vividus-framework/vividus/actions/runs/${GITHUB_RUN_ID} \
+                                                  -Pvividus.allure.executor.report-name="SauceLabs iOS Proxy tests report" \
+                                                  -Pvividus.saucelabs.sauce-connect.command-line-arguments="--proxy-localhost --no-ssl-bump-domains example.com" \
+                                                  -Pvividus.bdd.variables.global.app-url=sauce-storage:vividus-test-app-ios.zip \
+                                                  -Pvividus.statistics.print-failures=true
```
Comment (valfirst) [inline .github/workflows/gradle.yml:304]:
```suggestion
                                                  -Pvividus.bdd.variables.global.app-url=sauce-storage:vividus-test-app-ios.zip
```

File: .github/workflows/gradle.yml [hunk 8]
```diff
@@ -317,6 +344,26 @@ jobs:
...
+                                                  -Pvividus.allure.executor.build-order=${GITHUB_RUN_ID} \
+                                                  -Pvividus.allure.executor.build-name="SauceLabs Android Proxy Tests ${GITHUB_RUN_ID}" \
+                                                  -Pvividus.allure.executor.build-url=https://github.com/vividus-framework/vividus/actions/runs/${GITHUB_RUN_ID} \
+                                                  -Pvividus.allure.executor.report-url=https://github.com/vividus-framework/vividus/actions/runs/${GITHUB_RUN_ID} \
+                                                  -Pvividus.allure.executor.report-name="SauceLabs Android Proxy tests report" \
+                                                  -Pvividus.saucelabs.sauce-connect.command-line-arguments="--proxy-localhost --no-ssl-bump-domains example.com" \
+                                                  -Pvividus.bdd.variables.global.app-url=sauce-storage:vividus-test-app-android.zip \
+                                                  -Pvividus.statistics.print-failures=true
```
Comment (valfirst) [inline .github/workflows/gradle.yml:366]:
```suggestion
                                                  -Pvividus.bdd.variables.global.app-url=sauce-storage:vividus-test-app-android.zip
```

File: vividus-extension-selenium/build.gradle [hunk 9]
```diff
@@ -3,6 +3,7 @@ project.description = 'Vividus extension for selenium'
  ext {
      seleniumVersion = '4.1.1'
      browserupProxyVersion = '2.1.2'
+     nettyVersion = '4.1.70.Final'
```
Comment (valfirst) [inline vividus-extension-selenium/build.gradle:6]: https://github.com/vividus-framework/vividus/blob/master/vividus-plugin-web-app/build.gradle#L5

Comment (valfirst) [reply vividus-extension-selenium/build.gradle:6]: also should the same dependencies be removed from `vividus-plugin-web-app`?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: .github/workflows/gradle.yml [hunk 10]
```diff
@@ -282,6 +282,25 @@ jobs:
                                                   -Pvividus.allure.executor.report-url=https://github.com/vividus-framework/vividus/actions/runs/${GITHUB_RUN_ID} \
                                                   -Pvividus.allure.executor.report-name="SauceLabs iOS System tests report" \
                                                   -Pvividus.bdd.variables.global.app-url=sauce-storage:vividus-test-app-ios.zip
+            ./gradlew :vividus-tests:debugStories -Pvividus.configuration.environments=system/saucelabs/mobile_app/ios \
+                                                  -Pvividus.configuration.suites= \
+                                                  -Pvividus.configuration.profiles=saucelabs/mobile_app,mobile_app/ios \
+                                                  -Pvividus.bdd.story-loader.batch-1.resource-location=story/integration \
```
Comment (ikalinin1) [inline .github/workflows/gradle.yml:288]: the story location is incorrect

File: .github/workflows/gradle.yml [hunk 11]
```diff
@@ -317,6 +343,25 @@ jobs:
                                                   -Pvividus.allure.executor.report-url=https://github.com/vividus-framework/vividus/actions/runs/${GITHUB_RUN_ID} \
                                                   -Pvividus.allure.executor.report-name="SauceLabs Android System tests report" \
                                                   -Pvividus.bdd.variables.global.app-url=sauce-storage:vividus-test-app-android.zip
+            ./gradlew :vividus-tests:debugStories -Pvividus.configuration.environments=system/saucelabs/mobile_app/android \
+                                                  -Pvividus.configuration.suites= \
+                                                  -Pvividus.configuration.profiles=saucelabs/mobile_app,mobile_app/android \
+                                                  -Pvividus.bdd.story-loader.batch-1.resource-location=story/integration \
```
Comment (ikalinin1) [inline .github/workflows/gradle.yml:349]: https://github.com/vividus-framework/vividus/pull/2256/files#r773914257

File: vividus-extension-selenium/build.gradle (hunk 9 above)
Comment (valfirst) [inline vividus-extension-selenium/build.gradle:6]:
```suggestion
     nettyVersion = '4.1.72.Final'
```

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

File: .github/workflows/gradle.yml [hunk 12]
```diff
@@ -328,6 +366,44 @@ jobs:
...
+        else
+            echo No SAUCELABS_USER and/or SAUCELABS_KEY variables, System tests will be skipped
+        fi
+
+    - name: SauceLabs - Android Proxy - Mobile App - Publish system tests report
+      if: matrix.platform == 'ubuntu-latest' && always()
+      uses: actions/upload-artifact@v2
+      with:
+        name: SauceLabs - System Tests With Proxy- Allure report - Android
+        path: vividus-tests/output/reports/sl-andorid-proxy-allure
```
Comment (ikalinin1) [inline .github/workflows/gradle.yml:405]:
```suggestion
        path: vividus-tests/output/reports/sl-android-proxy-allure
```

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**